### DISCUSSION TITLE
Preserve formatting

### DISCUSSION THREAD

**Discussion by mightyiam** [kivikakk/comrak#354] [Q&A] [answered]
I'd like to parse Markdown, mutate the AST and then obtain a Markdown string again, preserving original formatting. Is that possible, please?

**Comment by kivikakk** [comment, marked as answer]
Not with Comrak — it doesn't preserve the original formatting details in its AST (per upstream [cmark](https://github.com/commonmark/cmark)). So, e.g. `_a_` will instead roundtrip to `*a*`; the output Markdown [when using CommonMark output](https://docs.rs/comrak/latest/comrak/fn.format_commonmark.html) will preserve the _meaning_ of the input, but not any precise formatting.

It is guaranteed to roundtrip stably, however; i.e. given some input A, when converted back to Markdown you'll get A' (which may or may not be equal to A); when converting A' you will again get A'.
